### PR TITLE
Add `UpperCamelCase` `EnumNamingStrategy`

### PR DESCRIPTION
**Pull request by lbenedetto**
Changes EnumNamingStrategies to provide all the same naming strategies as PropertyNamingStrategies. It accomplishes this by first normalizing the enum name to lowerCamelCase and then delegating to the various PropertyNamingStrategies implementations

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java [hunk 1]
```diff
@@ -9,18 +9,53 @@ public class EnumNamingStrategies {
 
     private EnumNamingStrategies() {}
 
+    /**
+     * Words other than first are capitalized and no separator is used between words.
+     * See {@link EnumNamingStrategies.LowerCamelCaseStrategy} for details.
+     *<p>
+     * Example external property names would be "numberValue", "namingStrategy", "theDefiniteProof".
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:16]: Add `@since 2.18`

**Comment by cowtowncoder** [comment]
Sounds good. Timing-wise this might not make it in 2.18 (hence refs to 2.19 are actually good).

Regardless, one thing we'd need if not yet done earlier is CLA. It's from

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and is usually easiest to do by printing, filling & signing, scan/photo, email to `cla` at fasterxml dot com. This only needs to be done once before the first contribution (in case I have asked for one earlier and forgot; if so, just LMK and I'll check archives).

**Comment by cowtowncoder** [comment]
Ok, CI failures for JDK 8, couple of things need to be fixed (no `var` in Java 8, `@Deprecated` doesn't have "forRemoval")

File: src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java [hunk 2]
```diff
@@ -43,87 +134,338 @@ private EnumNamingStrategies() {}
...
+        public String convertEnumToExternalName(String enumName) {
+            return PropertyNamingStrategies.LowerCaseStrategy.INSTANCE.translate(toBeanName(enumName));
         }
+    }
 
-        private static String toLowerCase(String string) {
-            int length = string.length();
-            StringBuilder builder = new StringBuilder(length);
-            for (int i = 0; i < length; i++) {
-                builder.append(charToLowerCaseIfUpper(string.charAt(i)));
```
Comment (lbenedetto) [inline src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:116]: Previous implementation had this strange way of converting to lowercase, but I noticed that the PropertyNamingStrategy implementations are perfectly happy to just use the standard toLowerCase function.

The two approaches handle special characters differently, but I don't think that applies to enum names.

File: src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java [hunk 3]
```diff
@@ -43,87 +134,338 @@ private EnumNamingStrategies() {}
...
+        StringBuilder out = null;
+        int iterationCnt = 0;
+        int lastSeparatorIdx = -1;
+
+        do {
+            lastSeparatorIdx = nextIndexOfUnderscore(enumName, lastSeparatorIdx + 1);
+            if (lastSeparatorIdx != -1) {
+                if (iterationCnt == 0) {
+                    out = new StringBuilder(enumName.length() + 4 * UNDERSCORE.length());
+                    out.append(enumName.substring(iterationCnt, lastSeparatorIdx).toLowerCase());
```
Comment (yawkat) [inline src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:436]: Please always pass a `Locale` to `toLowerCase` or there will be locale-specific issues.

Comment (lbenedetto) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:436]: See `LowerCaseStrategy` and `UpperSnakeCaseStrategy` in `PropertyNamingStrategies`. 
They use toLowerCase() and toUpperCase() without passing in the Locale, which is why I thought I could do the same here.

What Locale should I be passing? Should I update those strategies as well?

File: src/main/java/com/fasterxml/jackson/databind/PropertyNamingStrategies.java [hunk 4]
```diff
@@ -364,7 +364,7 @@ public String translate(String input) {
      * Conversion from internal name like "someOtherValue" would be into external name
      * if "someothervalue".
      */
-    public static class LowerCaseStrategy extends NamingBase
+    public static class  LowerCaseStrategy extends NamingBase
```
Comment (pjfanning) [inline src/main/java/com/fasterxml/jackson/databind/PropertyNamingStrategies.java:367]: can you revert this change?

File: src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java (hunk 3 above)
Comment (yawkat) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:436]: `Locale.ROOT` is the safe choice. imo the other strategies should use the same

@pjfanning @cowtowncoder WDYT

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:436]: I am not sure; I don't actually like the idea of case change being contextual since that can lead to interoperability problems; similar to using local timezone for date/time values.
But I am not 100% sure how others deal with this.

So how about this: in 2.x, follow existing (not best) Practice of not passing `Locale`.
And then file follow-up issue to change this in 3.0 (`master`) for all casing changes in `jackson-databind`.

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:436]: Actually, reading `Locale` Javadocs, I agree with @yawkat  -- `Locale.ROOT` would make sense.

Any changes to classes outside `EnumNamingStrategy` should go in separate PR tho, fwtw. But assuming safe (`Locale.ROOT` is defaulted to anyway?) could still go in 2.18.

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:436]: I'll change it -- and no, default is "local Locale" otherwise.

But one big challenge if changing to configurable: there's no access to configuration from `EnumNamingStrategy.convertEnumToExternalName()` -- unlike `PropertyNamingStrategy` which takes `MapperConfig`
Should change that for 3.0; will create an issue.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
Ok LGTM; will merge to get it out of the way.

Thank you @lbenedetto !

File: src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java [hunk 5]
```diff
@@ -43,87 +135,338 @@ private EnumNamingStrategies() {}
...
-                    }
-                    iterationCnt = lastSeparatorIdx + UNDERSCORE.length();
-                }
-            } while (lastSeparatorIdx != -1);
+    /**
+     * <p>
+     * An implementation of {@link EnumNamingStrategy} that converts enum names in the typical upper
+     * snake case format to upper camel case format.
+     * This implementation first normalizes to lower camel case using (see {@link LowerCamelCaseStrategy} for details)
+     * and then uses {@link PropertyNamingStrategies.UpperCamelCaseStrategy} to finish converting the name.
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:163]: Uhhh. BAD idea. I wish I had noticed before approving -- should not plug in to internal details of another class.

Just noticed when trying to merge into 3.0 (`master`).

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:163]: I will need to revert this, actually.

**Comment by cowtowncoder** [comment]
Ok my apologies: I should have noticed one thing -- implements should NOT use internal implementation of `PropertyNamingStrategy` / `-Strategies`: that's not an extension point.

File: src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java (hunk 5 above)
Comment (lbenedetto) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:163]: Sorry, I thought it made sense to reuse this preexisting battle-tested implementation for converting lowerCamelCase into other specific cases.

Do you have a recommendation for how this should be achieved instead? Access the same code in a different way? Implement my own logic from scratch? Copy paste?

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:163]: Right, I agree with the intent here, just not the mechanism.

Not 100% sure of a good way (can't think of a great way at least): copy pasting (with a comment pointing to source) might be ok here, given it shouldn't change a lot.
Otherwise a util class would be good. Unfortunately there isn't any obvious choice... closest might be `BeanUtil`, if method(s) in question can be static.

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:163]: ... also, adding something like `NamingUtil` in `src/main/java/com/fasterxml/jackson/databind/util` would be ok if that makes more sense. It's bit dubious given only 2 use cases but... maybe that's the way to do it.

Apologies for catching this bit late.

Comment (lbenedetto) [reply src/main/java/com/fasterxml/jackson/databind/EnumNamingStrategies.java:163]: Here is the updated pull request:
https://github.com/FasterXML/jackson-databind/pull/4728
